# Instructions for 

## Massively parallel reporter assay with tail-length sequencing 


## Part 1: for random-sequence libraries

`MPRA_tail_seq_HMM_v2.4.py`

### Options for the input
* `-o`: Prefix for the output file
* `-q`: Input fastq file
* `-i`: Input intensity file
* `-s`: Input file for calculated "A signals"
* `-c`: If specified, the constant regionn will be examined (relevant sequences and parameters can be changed in the variable `params`)
* `-a`: If specified, the 3' adapter will examined (relevant sequences and parameters can be changed in the variable `params`)
* `-p`: Starting position for the poly(A) region (1-based)
* `-sp`: If specified, the spike-in sequences will be examined (relevant sequences and parameters can be changed in the variable `params`)

Examples of running the script

```
python MPRA_tail_seq_HMM_v2.4.py -c -a -sp -q input.fq.gz -i input_intensity.txt.gz -o output -p 82

```
or with calculated "A signals"
```
python MPRA_tail_seq_HMM_v2.4.py -s A_signals.txt.gz -o output

```


## Part 2: for single-nucleotide mutagenesis libraries

`MPRA_tail_seq_ref_HMM_v1.3.py`

### Options for the input
* `-b`: Input bam file, generated by aligning fastq files to a reference
* `-fq`: Input fastq file that contains the poly(A) region
* `-u`: Input fastq file that contains the UMI 
* `-i`: Input intensity file
* `-s`: Input file for calculated "A signals"
* `-m`: A pre-definded HMM model file in `xml` format (optional)
* `-ps`: Starting position for the poly(A) region (1-based, optional)
* `-pm`: The maximal length of a poly(A) tail (optional)
* `-a`: If specified, annotate each mapped read 
* `-o`: Prefix for the output file

Examples of running the script
```
python MPRA_tail_seq_ref_HMM_v1.3.py -a -b reads_mapped.bam -fq input.fq.gz -u umi.fq.gz -i input_intensity.txt.gz -o output -ps 1 -pm 200

```
